File: audit/deepmind/trace_learner.lua
#!/usr/bin/env luajit

local ok_torch, torch = pcall(require, "torch")
if not ok_torch then error("torch is required") end
local ok_nn, nn = pcall(require, "nn")
if not ok_nn then error("nn is required") end
require("initenv")
local ffi = require("ffi")

local function getenv(name, default)
  local value = os.getenv(name)
  if value == nil or value == "" then return default end
  return value
end

local OUT_ROOT = getenv("OUT", "audit_outputs/stage5_learner")
local FIXTURE_DIR = getenv("STAGE5_FIXTURE_DIR", OUT_ROOT .. "/learner_fixture")
local MODEL_DIR = getenv("STAGE5_MODEL_DIR", OUT_ROOT .. "/model_exchange")
local OUT = getenv("DEEPMIND_LEARNER_OUT", OUT_ROOT .. "/deepmind_learner.jsonl")
local BATCH_SIZE = tonumber(getenv("STAGE5_BATCH_SIZE", "32"))
local LR = tonumber(getenv("LEARNER_LR", "0.00025"))
local GAMMA = tonumber(getenv("GAMMA", "0.99"))
local ACTION_COUNT = tonumber(getenv("ACTION_COUNT", "4"))
local HIST_LEN = 4
local FRAME_H = 84
local FRAME_W = 84

local function shell_quote(path) return "'" .. tostring(path):gsub("'", "'\\''") .. "'" end
local function mkdir_p(path) os.execute("mkdir -p " .. shell_quote(path)) end
local function dirname(path) return tostring(path):match("^(.*)/[^/]*$") or "." end

local function json_escape(value)
  value = tostring(value):gsub("\\", "\\\\"):gsub('"', '\\"')
  value = value:gsub("\n", "\\n"):gsub("\r", "\\r"):gsub("\t", "\\t")
  return value
end

local function is_array(tbl)
  if type(tbl) ~= "table" then return false end
  local count, max_key = 0, 0
  for key, _ in pairs(tbl) do
    if type(key) ~= "number" then return false end
    if key > max_key then max_key = key end
    count = count + 1
  end
  return max_key == count
end

local function sorted_keys(tbl)
  local keys = {}
  for key, _ in pairs(tbl) do keys[#keys + 1] = key end
  table.sort(keys, function(a, b) return tostring(a) < tostring(b) end)
  return keys
end

local function encode_json(value)
  local value_type = type(value)
  if value_type == "nil" then return "null"
  elseif value_type == "boolean" then return value and "true" or "false"
  elseif value_type == "number" then return tostring(value)
  elseif value_type == "string" then return '"' .. json_escape(value) .. '"'
  elseif value_type == "table" then
    if is_array(value) then
      local parts = {}
      for i = 1, #value do parts[#parts + 1] = encode_json(value[i]) end
      return "[" .. table.concat(parts, ",") .. "]"
    end
    local parts = {}
    for _, key in ipairs(sorted_keys(value)) do
      if value[key] ~= nil then parts[#parts + 1] = '"' .. json_escape(key) .. '":' .. encode_json(value[key]) end
    end
    return "{" .. table.concat(parts, ",") .. "}"
  end
  return '"' .. json_escape(tostring(value)) .. '"'
end

local function tensor_shape(tensor)
  local shape = {}
  for dim = 1, tensor:dim() do shape[#shape + 1] = tensor:size(dim) end
  return shape
end

local function sha256_file(path)
  local proc = assert(io.popen("sha256sum " .. shell_quote(path)))
  local line = proc:read("*l")
  proc:close()
  return line:match("^(%w+)")
end

local float_box = ffi.new("float[1]")
local float_bytes = ffi.cast("uint8_t*", float_box)

local function read_float32_raw(path, shape)
  local handle = assert(io.open(path, "rb"))
  local bytes = handle:read("*a")
  handle:close()
  local expected = 1
  for _, dim in ipairs(shape) do expected = expected * dim end
  assert(#bytes == expected * 4, "bad float32 byte count for " .. path)
  local tensor = torch.FloatTensor(unpack(shape))
  local flat = tensor:view(tensor:nElement())
  local byte_ptr = ffi.cast("const uint8_t*", bytes)
  for i = 1, expected do
    float_bytes[0] = byte_ptr[(i - 1) * 4]
    float_bytes[1] = byte_ptr[(i - 1) * 4 + 1]
    float_bytes[2] = byte_ptr[(i - 1) * 4 + 2]
    float_bytes[3] = byte_ptr[(i - 1) * 4 + 3]
    flat[i] = tonumber(float_box[0])
  end
  return tensor:contiguous()
end

local function read_uint8_tensor(path, shape)
  local handle = assert(io.open(path, "rb"))
  local bytes = handle:read("*a")
  handle:close()
  local expected = 1
  for _, dim in ipairs(shape) do expected = expected * dim end
  assert(#bytes == expected, "bad uint8 byte count for " .. path)
  local tensor = torch.ByteTensor(unpack(shape))
  local flat = tensor:view(tensor:nElement())
  for i = 1, expected do flat[i] = string.byte(bytes, i) end
  return tensor:contiguous()
end

local function read_number_list(path, kind)
  local values = {}
  local handle = assert(io.open(path, "r"))
  for line in handle:lines() do
    local stripped = line:match("^%s*(.-)%s*$")
    if stripped ~= "" then values[#values + 1] = tonumber(stripped) end
  end
  handle:close()
  local tensor
  if kind == "long" then tensor = torch.LongTensor(#values)
  elseif kind == "byte" then tensor = torch.ByteTensor(#values)
  else tensor = torch.FloatTensor(#values) end
  for i, value in ipairs(values) do tensor[i] = value end
  return tensor
end

local function load_model()
  local create_network = require("convnet_atari3")
  local net = create_network({
    hist_len = HIST_LEN,
    ncols = 1,
    input_dims = {HIST_LEN, FRAME_H, FRAME_W},
    n_actions = ACTION_COUNT,
    gpu = -1,
    verbose = 0,
  })
  net:float()
  local defs = {
    {name = "conv1", module_index = 2, weight_shape = {32, 4, 8, 8}, bias_shape = {32}},
    {name = "conv2", module_index = 4, weight_shape = {64, 32, 4, 4}, bias_shape = {64}},
    {name = "conv3", module_index = 6, weight_shape = {64, 64, 3, 3}, bias_shape = {64}},
    {name = "fc1", module_index = 9, weight_shape = {512, 3136}, bias_shape = {512}},
    {name = "fc2", module_index = 11, weight_shape = {ACTION_COUNT, 512}, bias_shape = {ACTION_COUNT}},
  }
  for _, def in ipairs(defs) do
    local module = net.modules[def.module_index]
    module.weight:copy(read_float32_raw(MODEL_DIR .. "/" .. def.name .. "_weight.float32", def.weight_shape))
    module.bias:copy(read_float32_raw(MODEL_DIR .. "/" .. def.name .. "_bias.float32", def.bias_shape))
  end
  return net, defs
end

local function exchange_layer_contract(defs)
  local rows = {}
  for _, def in ipairs(defs) do
    rows[#rows + 1] = {
      name = def.name,
      weight_shape = def.weight_shape,
      bias_shape = def.bias_shape,
      weight_sha256 = sha256_file(MODEL_DIR .. "/" .. def.name .. "_weight.float32"),
      bias_sha256 = sha256_file(MODEL_DIR .. "/" .. def.name .. "_bias.float32"),
    }
  end
  return rows
end

local function matrix_rows(tensor, rows)
  local out = {}
  local n = math.min(rows or 8, tensor:size(1))
  for i = 1, n do
    out[i] = {}
    for j = 1, tensor:size(2) do out[i][j] = tonumber(tensor[i][j]) end
  end
  return out
end

local function tensor_list(tensor, limit)
  local flat = tensor:float():contiguous():view(tensor:nElement())
  local n = flat:nElement()
  if limit and limit < n then n = limit end
  local out = {}
  for i = 1, n do out[i] = tonumber(flat[i]) end
  return out
end

local function tensor_summary(tensor)
  local data = tensor:float():contiguous()
  return {
    shape = tensor_shape(data),
    dtype = "float32",
    hash = "not_compared",
    mean = tonumber(data:mean()),
    max_abs = tonumber(data:abs():max()),
    l2 = tonumber(torch.sqrt(torch.pow(data:double(), 2):sum())),
    first_values = tensor_list(data, 8),
  }
end

local function layer_summaries(net, include_grads)
  local defs = {
    {name = "conv.0.weight", module_index = 2, field = "weight"},
    {name = "conv.0.bias", module_index = 2, field = "bias"},
    {name = "conv.2.weight", module_index = 4, field = "weight"},
    {name = "conv.2.bias", module_index = 4, field = "bias"},
    {name = "conv.4.weight", module_index = 6, field = "weight"},
    {name = "conv.4.bias", module_index = 6, field = "bias"},
    {name = "fc.0.weight", module_index = 9, field = "weight"},
    {name = "fc.0.bias", module_index = 9, field = "bias"},
    {name = "fc.2.weight", module_index = 11, field = "weight"},
    {name = "fc.2.bias", module_index = 11, field = "bias"},
  }
  local rows = {}
  for _, def in ipairs(defs) do
    local module = net.modules[def.module_index]
    local item = {name = def.name, param = tensor_summary(module[def.field])}
    if include_grads then
      local grad_field = def.field == "weight" and "gradWeight" or "gradBias"
      item.grad = tensor_summary(module[grad_field])
    end
    rows[#rows + 1] = item
  end
  return rows
end

local states_uint8 = read_uint8_tensor(FIXTURE_DIR .. "/states_uint8.bin", {BATCH_SIZE, HIST_LEN, FRAME_H, FRAME_W})
local next_states_uint8 = read_uint8_tensor(FIXTURE_DIR .. "/next_states_uint8.bin", {BATCH_SIZE, HIST_LEN, FRAME_H, FRAME_W})
local s = states_uint8:float():div(255)
local s2 = next_states_uint8:float():div(255)
local a = read_number_list(FIXTURE_DIR .. "/actions_one_based.txt", "long")
local a_zero = read_number_list(FIXTURE_DIR .. "/actions_zero_based.txt", "long")
local r = read_number_list(FIXTURE_DIR .. "/rewards.txt", "float")
local term = read_number_list(FIXTURE_DIR .. "/terminals.txt", "byte")

local net, defs = load_model()
local target_net = net:clone()

local rows = {}
rows[#rows + 1] = {
  phase = "model_contract",
  source = "deepmind",
  step = 0,
  architecture = "convnet_atari3_compatible",
  qnetwork_source_kind = "deepmind_convnet_atari3",
  loss_source_kind = "NeuralQLearner_getQUpdate_and_network_backward",
  optimizer_source_kind = "NeuralQLearner_qLearnMinibatch_manual_centered_rmsprop",
  action_count = ACTION_COUNT,
  input_shape = {4, 84, 84},
  parameter_layers = exchange_layer_contract(defs),
}

local q_values = net:forward(s):float()
local next_q_values = target_net:forward(s2):float()
local max_next_q = next_q_values:max(2):view(BATCH_SIZE)
rows[#rows + 1] = {
  phase = "forward",
  source = "deepmind",
  step = 1,
  input_contract = "effective_float32_0_1_network_input",
  batch_size = BATCH_SIZE,
  q_values_first_rows = matrix_rows(q_values, 8),
  next_q_values_first_rows = matrix_rows(next_q_values, 8),
  q_values_summary = tensor_summary(q_values),
  next_q_values_summary = tensor_summary(next_q_values),
}

local term_mask = term:clone():float():mul(-1):add(1)
local q2 = max_next_q:clone():mul(GAMMA):cmul(term_mask)
local target_values = r:clone():float():add(q2)
local q_selected = torch.FloatTensor(BATCH_SIZE)
for i = 1, BATCH_SIZE do q_selected[i] = q_values[i][a[i]] end
local delta_unclipped = target_values:clone():add(-1, q_selected)
local clipped_delta = delta_unclipped:clone()
clipped_delta[clipped_delta:ge(1)] = 1
clipped_delta[clipped_delta:le(-1)] = -1
local sparse_targets = torch.zeros(BATCH_SIZE, ACTION_COUNT):float()
for i = 1, BATCH_SIZE do sparse_targets[i][a[i]] = clipped_delta[i] end

rows[#rows + 1] = {
  phase = "bellman_target",
  source = "deepmind",
  step = 2,
  gamma = GAMMA,
  actions_zero_based = tensor_list(a_zero),
  actions_one_based = tensor_list(a),
  rewards = tensor_list(r),
  terminals = tensor_list(term),
  q_selected = tensor_list(q_selected),
  max_next_q = tensor_list(max_next_q),
  target_values = tensor_list(target_values),
  delta_unclipped = tensor_list(delta_unclipped),
  clipped_delta = tensor_list(clipped_delta),
}

local w, dw = net:getParameters()
dw:zero()
net:backward(s, sparse_targets)
local grad_norm = tonumber(torch.sqrt(torch.pow(dw:double(), 2):sum()))

rows[#rows + 1] = {
  phase = "loss_gradient_contract",
  source = "deepmind",
  step = 3,
  loss_mode = "deepmind_sparse_clipped_delta_backward_no_scalar_loss",
  scalar_loss = "none",
  output_gradient_scale = "d_network_output_is_sparse_clipped_delta_no_batch_mean",
  grad_norm = grad_norm,
  pred_minus_target_first_values = tensor_list(q_selected:clone():add(-1, target_values), 8),
}
rows[#rows + 1] = {
  phase = "gradient_summary",
  source = "deepmind",
  step = 4,
  parameter_layers = layer_summaries(net, true),
}

local before = w:clone()
local g = dw:clone():fill(0)
local g2 = dw:clone():fill(0)
local tmp = dw:clone():fill(0)
local deltas = dw:clone():fill(0)
g:mul(0.95):add(0.05, dw)
tmp:cmul(dw, dw)
g2:mul(0.95):add(0.05, tmp)
tmp:cmul(g, g)
tmp:mul(-1)
tmp:add(g2)
tmp:add(0.01)
tmp:sqrt()
deltas:mul(0):addcdiv(LR, dw, tmp)
w:add(deltas)
local param_delta = w:clone():add(-1, before)

rows[#rows + 1] = {
  phase = "optimizer_update",
  source = "deepmind",
  step = 5,
  optimizer_mode = "NeuralQLearner_manual_centered_rmsprop_on_sparse_clipped_delta_gradients",
  lr = LR,
  alpha = 0.95,
  epsilon = 0.01,
  epsilon_placement = "inside_sqrt",
  flat_parameter_delta = tensor_summary(param_delta),
  post_update_layers = layer_summaries(net, false),
}

mkdir_p(dirname(OUT))
local out = assert(io.open(OUT, "w"))
for _, row in ipairs(rows) do
  out:write(encode_json(row))
  out:write("\n")
end
out:close()
print("wrote " .. OUT)
print("rows: " .. tostring(#rows))

File: audit/pytorch/trace_learner.py
#!/usr/bin/env python
from __future__ import annotations

import argparse
import importlib.util
import json
import os
import sys
from pathlib import Path
from typing import Any, Callable

import numpy as np
import torch
import torch.nn as nn

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from common import arr_hash, canonical_float, write_jsonl  # noqa: E402
from stage5_common import load_stage5_manifest, read_raw_array  # noqa: E402


class FallbackQNetwork(nn.Module):
    def __init__(self, num_actions: int):
        super().__init__()
        self.conv = nn.Sequential(
            nn.Conv2d(4, 32, kernel_size=8, stride=4),
            nn.ReLU(),
            nn.Conv2d(32, 64, kernel_size=4, stride=2),
            nn.ReLU(),
            nn.Conv2d(64, 64, kernel_size=3, stride=1),
            nn.ReLU(),
        )
        self.fc = nn.Sequential(nn.Linear(64 * 7 * 7, 512), nn.ReLU(), nn.Linear(512, num_actions))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = x / 255.0
        x = self.conv(x)
        x = torch.flatten(x, start_dim=1)
        return self.fc(x)


def parse_args() -> argparse.Namespace:
    out = os.getenv("OUT", "audit_outputs/stage5_learner")
    atari_dir = os.getenv("ATARI_DIR", str(Path.cwd()))
    parser = argparse.ArgumentParser(description="Trace PyTorch Atari learner math on the frozen Stage 5 minibatch.")
    parser.add_argument("--fixture-dir", default=os.path.join(out, "learner_fixture"))
    parser.add_argument("--model", default=os.path.join(out, "model_exchange", "pytorch_model.pt"))
    parser.add_argument("--atari-dir", default=atari_dir)
    parser.add_argument("--out", default=os.path.join(out, "pytorch_learner.jsonl"))
    parser.add_argument("--lr", type=float, default=float(os.getenv("LEARNER_LR", "0.00025")))
    parser.add_argument("--gamma", type=float, default=float(os.getenv("GAMMA", "0.99")))
    return parser.parse_args()


def import_from_path(module_name: str, path: Path) -> Any:
    spec = importlib.util.spec_from_file_location(module_name, path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"could not import {path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def load_qnetwork_class(atari_dir: str | Path) -> tuple[type[nn.Module], str]:
    network_path = Path(atari_dir) / "network.py"
    if not network_path.exists():
        return FallbackQNetwork, "fallback_audit_qnetwork"
    module = import_from_path("stage5_atari_network_trace", network_path)
    return module.QNetwork, str(network_path)


def load_loss(atari_dir: str | Path) -> tuple[Callable[[torch.Tensor, torch.Tensor], torch.Tensor], str]:
    train_path = Path(atari_dir) / "train_nature.py"
    if train_path.exists():
        try:
            module = import_from_path("stage5_train_nature", train_path)
            return module.clipped_td_error_loss, str(train_path)
        except Exception:
            pass

    def fallback_loss(predictions: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
        td_errors = predictions - targets
        abs_errors = td_errors.abs()
        quadratic = torch.minimum(abs_errors, torch.ones_like(abs_errors))
        linear = abs_errors - quadratic
        return (0.5 * quadratic.pow(2) + linear).mean()

    return fallback_loss, "fallback_mean_huber"


def load_optimizer_class(atari_dir: str | Path) -> tuple[type[torch.optim.Optimizer], str]:
    opt_path = Path(atari_dir) / "deepmind_rmsprop.py"
    if not opt_path.exists():
        raise RuntimeError(f"missing DeepMindRMSprop implementation: {opt_path}")
    module = import_from_path("stage5_deepmind_rmsprop", opt_path)
    return module.DeepMindRMSprop, str(opt_path)


def to_float_list(array: np.ndarray | torch.Tensor, limit: int | None = None) -> list[Any]:
    if isinstance(array, torch.Tensor):
        data = array.detach().cpu().numpy()
    else:
        data = np.asarray(array)
    flat = data.reshape(-1)
    if limit is not None:
        flat = flat[:limit]
    return [canonical_float(float(value), digits=10) for value in flat.tolist()]


def matrix_rows(tensor: torch.Tensor, rows: int = 8) -> list[list[Any]]:
    data = tensor.detach().cpu().numpy()
    return [[canonical_float(float(value), digits=10) for value in row] for row in data[:rows].tolist()]


def summarize_array(array: np.ndarray | torch.Tensor) -> dict[str, Any]:
    if isinstance(array, torch.Tensor):
        data = array.detach().cpu().numpy()
    else:
        data = np.asarray(array)
    data = np.ascontiguousarray(data.astype(np.float32, copy=False))
    return {
        "shape": [int(dim) for dim in data.shape],
        "dtype": str(data.dtype),
        "hash": arr_hash(data),
        "mean": canonical_float(float(data.mean()), digits=10),
        "max_abs": canonical_float(float(np.abs(data).max()), digits=10),
        "l2": canonical_float(float(np.sqrt(np.square(data.astype(np.float64)).sum())), digits=10),
        "first_values": to_float_list(data, limit=8),
    }


def layer_summaries(model: nn.Module, include_grads: bool = False) -> list[dict[str, Any]]:
    rows = []
    for name, parameter in model.named_parameters():
        item = {"name": name, "param": summarize_array(parameter)}
        if include_grads:
            item["grad"] = None if parameter.grad is None else summarize_array(parameter.grad)
        rows.append(item)
    return rows


def load_fixture(fixture_dir: Path) -> dict[str, np.ndarray]:
    manifest = load_stage5_manifest(fixture_dir)
    tensors = {entry["name"]: entry for entry in manifest["tensors"]}
    return {
        name: read_raw_array(fixture_dir / entry["path"], entry["shape"], entry["dtype"])
        for name, entry in tensors.items()
    }


def exchange_layer_contract(checkpoint: dict[str, Any]) -> list[dict[str, Any]]:
    rows = []
    for layer in checkpoint["deepmind_manifest"]["layers"]:
        rows.append(
            {
                "name": layer["name"],
                "weight_shape": [int(dim) for dim in layer["weight_shape"]],
                "bias_shape": [int(dim) for dim in layer["bias_shape"]],
                "weight_sha256": layer["weight_sha256"],
                "bias_sha256": layer["bias_sha256"],
            }
        )
    return rows


def main() -> None:
    args = parse_args()
    fixture_dir = Path(args.fixture_dir)
    manifest = load_stage5_manifest(fixture_dir)
    arrays = load_fixture(fixture_dir)
    checkpoint = torch.load(args.model, map_location="cpu")

    qnetwork_cls, qnetwork_source = load_qnetwork_class(args.atari_dir)
    loss_fn, loss_source = load_loss(args.atari_dir)
    optimizer_cls, optimizer_source = load_optimizer_class(args.atari_dir)

    action_count = int(checkpoint["deepmind_manifest"]["action_count"])
    q_net = qnetwork_cls(action_count)
    target_net = qnetwork_cls(action_count)
    q_net.load_state_dict(checkpoint["state_dict"])
    target_net.load_state_dict(checkpoint["state_dict"])
    q_net.train()
    target_net.eval()

    states = torch.as_tensor(arrays["states_uint8"], dtype=torch.float32)
    next_states = torch.as_tensor(arrays["next_states_uint8"], dtype=torch.float32)
    actions = torch.as_tensor(arrays["actions_zero_based"], dtype=torch.long)
    rewards = torch.as_tensor(arrays["rewards_float32"], dtype=torch.float32)
    terminals = torch.as_tensor(arrays["terminals_uint8"], dtype=torch.float32)

    rows: list[dict[str, Any]] = []
    rows.append(
        {
            "phase": "model_contract",
            "source": "pytorch",
            "step": 0,
            "architecture": "convnet_atari3_compatible",
            "qnetwork_source_kind": "real_atari_qnetwork" if "network.py" in qnetwork_source else "fallback",
            "loss_source_kind": "real_train_nature_loss" if "train_nature.py" in loss_source else "fallback",
            "optimizer_source_kind": "real_atari_deepmind_rmsprop" if "deepmind_rmsprop.py" in optimizer_source else "fallback",
            "action_count": action_count,
            "input_shape": [4, 84, 84],
            "parameter_layers": exchange_layer_contract(checkpoint),
        }
    )

    q_values = q_net(states)
    with torch.no_grad():
        next_q_values = target_net(next_states)
        max_next_q = next_q_values.max(dim=1).values
        target_values = rewards + (1.0 - terminals) * float(args.gamma) * max_next_q
    q_selected = q_values.gather(1, actions.unsqueeze(1)).squeeze(1)
    delta_unclipped = target_values - q_selected.detach()
    clipped_delta = torch.clamp(delta_unclipped, -1.0, 1.0)

    rows.append(
        {
            "phase": "forward",
            "source": "pytorch",
            "step": 1,
            "input_contract": "effective_float32_0_1_network_input",
            "batch_size": int(states.shape[0]),
            "q_values_first_rows": matrix_rows(q_values),
            "next_q_values_first_rows": matrix_rows(next_q_values),
            "q_values_summary": summarize_array(q_values),
            "next_q_values_summary": summarize_array(next_q_values),
        }
    )
    rows.append(
        {
            "phase": "bellman_target",
            "source": "pytorch",
            "step": 2,
            "gamma": canonical_float(float(args.gamma), digits=10),
            "actions_zero_based": [int(x) for x in arrays["actions_zero_based"].tolist()],
            "actions_one_based": [int(x) for x in arrays["actions_one_based"].tolist()],
            "rewards": to_float_list(rewards),
            "terminals": [int(x) for x in arrays["terminals_uint8"].tolist()],
            "q_selected": to_float_list(q_selected),
            "max_next_q": to_float_list(max_next_q),
            "target_values": to_float_list(target_values),
            "delta_unclipped": to_float_list(delta_unclipped),
            "clipped_delta": to_float_list(clipped_delta),
        }
    )

    predictions = q_selected
    loss = loss_fn(predictions, target_values.detach())
    optimizer = optimizer_cls(q_net.parameters(), lr=float(args.lr), alpha=0.95, eps=0.01)
    optimizer.zero_grad()
    loss.backward()
    grad_norm = torch.sqrt(
        sum(
            parameter.grad.detach().pow(2).sum()
            for parameter in q_net.parameters()
            if parameter.grad is not None
        )
    )

    rows.append(
        {
            "phase": "loss_gradient_contract",
            "source": "pytorch",
            "step": 3,
            "loss_mode": "pytorch_current_mean_huber_scalar_loss",
            "scalar_loss": canonical_float(float(loss.detach().item()), digits=10),
            "output_gradient_scale": "d_loss_d_q_selected_is_clamp(q-target,-1,1)/batch_size",
            "grad_norm": canonical_float(float(grad_norm.item()), digits=10),
            "pred_minus_target_first_values": to_float_list((predictions.detach() - target_values.detach()), limit=8),
        }
    )
    rows.append(
        {
            "phase": "gradient_summary",
            "source": "pytorch",
            "step": 4,
            "parameter_layers": layer_summaries(q_net, include_grads=True),
        }
    )

    before = {name: parameter.detach().clone() for name, parameter in q_net.named_parameters()}
    optimizer.step()
    deltas = []
    for name, parameter in q_net.named_parameters():
        delta = parameter.detach() - before[name]
        deltas.append({"name": name, "delta": summarize_array(delta)})

    rows.append(
        {
            "phase": "optimizer_update",
            "source": "pytorch",
            "step": 5,
            "optimizer_mode": "real_atari_deepmind_rmsprop_on_current_huber_gradients",
            "lr": canonical_float(float(args.lr), digits=10),
            "alpha": 0.95,
            "epsilon": 0.01,
            "epsilon_placement": "inside_sqrt",
            "parameter_deltas": deltas,
            "post_update_layers": layer_summaries(q_net, include_grads=False),
        }
    )

    write_jsonl(args.out, rows)
    print(f"wrote {args.out}")
    print(f"rows: {len(rows)}")
    print(f"fixture_content_hash: {manifest['content_hash']}")


if __name__ == "__main__":
    main()
